controller/volume_controller.go

```diff
@@ -2999,6 +3000,127 @@ func (vc *VolumeController) ResolveRefAndEnqueue(namespace string, ref *metav1.O
 	vc.enqueueVolume(volume)
 }
 
+// createNewRecurringJobNameUsed use a new name to create a new recurring job that original name exists but configuration is different
+func createNewRecurringJobNameUsed(log *logrus.Entry, ds *datastore.DataStore, job *longhorn.RecordVolumeRecurringJob) (string, error) {
+	newJobName := util.RandomID()
+	if _, err := ds.GetRecurringJob(newJobName); err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.WithError(err).Warnf("Failed to get a new recurring job name %v", newJobName)
+			return "", err
+		}
+		oldJobName := job.JobSpec.Name
+		job.JobSpec.Name = newJobName
+		log.Debugf("Create a new recurring job %v instead of a existing job name %v when restoring recurring jobs", newJobName, oldJobName)
+		if _, err := ds.CreateRecurringJob(&longhorn.RecurringJob{ObjectMeta: metav1.ObjectMeta{Name: newJobName}, Spec: job.JobSpec}); err != nil {
+			log.WithError(err).Warnf("Failed to create a new recurring job %v for restoring the job %v", newJobName, oldJobName)
+			return "", err
+		}
+	}
+	return newJobName, nil
+}
+
+// createRecurringJobNotExistAndLabel add recurring jobs and group back to a restoring volume and create recurring jobs if not exist
+func createRecurringJobNotExistAndLabel(log *logrus.Entry, ds *datastore.DataStore, v *longhorn.Volume, jobName string, job *longhorn.RecordVolumeRecurringJob) error {
+	if existJob, err := ds.GetRecurringJob(jobName); err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.WithError(err).Warnf("Failed to get the recurring job %v", jobName)
+			return err
+		}
+		// create the recurring job if not exist
+		log.Debugf("Create the recurring job %v when restoring recurring jobs", jobName)
+		if _, err := ds.CreateRecurringJob(&longhorn.RecurringJob{ObjectMeta: metav1.ObjectMeta{Name: jobName}, Spec: job.JobSpec}); err != nil {
+			log.WithError(err).Warnf("Failed to create a recurring job %v", jobName)
+			return err
+		}
+	} else if !reflect.DeepEqual(existJob.Spec, job.JobSpec) {
```

Comment:
What happens if the volume restores multiple times? Will this create a bunch of new recurring jobs?

Reply:
Yes, it will create a bunch of new recurring jobs.
Modified. Reuse the re-created recurring jobs restored from the backup target.

Reply:
How do you guarantee the function won't repeatedly restore a recurring job of which the name is occupied?

Reply:
It would search the recurring jobs having the name with the prefix `restored-recurring-job-` and compare the Spec before create a new recurring job with the prefix.